lib/core/src/Ouroboros/Network/Client/Wallet.hs

```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
+    clientStAcquiring qry oldPt era retries = LSQ.ClientStAcquiring
+        { recvMsgAcquired = clientStAcquired qry era
+        , recvMsgFailure = \_failure -> do
+            (newEra, newPt) <- second getTipPoint <$> (atomically $ do
+                t <- readTVar tipVar
+                guard (getTipPoint (snd t) /= oldPt)
+                return t)
+            pure $ LSQ.SendMsgAcquire newPt (clientStAcquiring qry newPt newEra (retries + 1))
         }
 
     clientStAcquired
-        :: forall state. Query block state
-        -> (LocalStateQueryResult state -> m ())
-        -> LSQ.ClientStAcquired block (Point block) (Query block) m Void
-    clientStAcquired query respond =
-        LSQ.SendMsgQuery query (clientStQuerying respond)
-
-    -- By re-acquiring rather releasing the state with 'MsgRelease' it
-    -- enables optimisations on the server side.
-    clientStAcquiredAgain
-        :: m (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
-    clientStAcquiredAgain = awaitNextCmd <&> \case
-        CmdQueryLocalState pt query respond ->
-            LSQ.SendMsgReAcquire pt (clientStAcquiring query respond)
-
-    clientStQuerying
-        :: forall state. (LocalStateQueryResult state -> m ())
-        -> LSQ.ClientStQuerying block (Point block) (Query block) m Void state
-    clientStQuerying respond = LSQ.ClientStQuerying
-        { recvMsgResult = \result -> do
-            respond (Right result)
-            clientStAcquiredAgain
-        }
+        :: LocalStateQueryCmd block m
+        -> AnyCardanoEra
+        -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
+    clientStAcquired (SomeLSQ cmd respond) era = go cmd $ \res -> do
+        LSQ.SendMsgRelease (respond res >> clientStIdle)
+        -- TODO: We might not want to release. Re-acquiring, or even staying
+        -- acquired should be more efficient better.
+      where
+          go
+              :: forall a. LSQ block m a
+              -> (a -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void))
+              -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
+          go (LSQPure a) cont = (cont a)
+          go (LSQry qry) cont = LSQ.SendMsgQuery qry $ LSQ.ClientStQuerying $ \res -> do
+              pure $ cont res
+          go (LSQBind ma f) cont = go ma $ \a -> do
+              go (f a) $ \b -> cont b
+          go (LSQEra) cont = cont era
 
     awaitNextCmd :: m (LocalStateQueryCmd block m)
     awaitNextCmd = atomically $ readTQueue queue
 
+-- Monad for composing local state queries for the node /tip/.
+data LSQ block (m :: * -> *) a where
+    LSQPure :: a -> LSQ block m a
+    LSQBind :: LSQ block m a -> (a -> LSQ block m b) -> LSQ block m b
+    LSQry :: (Query block res) -> LSQ block m res
+    LSQEra :: LSQ block m AnyCardanoEra
+
+instance Functor (LSQ block m) where
+    fmap = liftM
+
+instance Applicative (LSQ block m) where
+    pure  = LSQPure
+    (<*>) = ap
```

Comment:
The fact that `ap` is defined in terms of `Monad` is utterly confusing to me :smile: ... I guess this is because I started with PureScript and not Haskell, where everything comes from `Applicative` whereas Haskell usually defines things from `Monad` and downwards.